curriculum/challenges/english/20-upcoming-python/learn-classes-and-objects-by-building-a-sudoku-solver/6569fca3cd7a9f52f322a298.md

```diff
@@ -0,0 +1,110 @@
+assert.match(code, /print\(gameboard\)/);
+```
+
+# --seed--
+
+## --seed-contents--
+
+```py
+class Board:
+    def __init__(self, board):
+        self.board = board
+
+    def __str__(self):
+        upper_lines = f'\n╔═══{"╤═══"*2}{"╦═══"}{"╤═══"*2}{"╦═══"}{"╤═══"*2}╗\n'
+        middle_lines = f'╟───{"┼───"*2}{"╫───"}{"┼───"*2}{"╫───"}{"┼───"*2}╢\n'
+        lower_lines = f'╚═══{"╧═══"*2}{"╩═══"}{"╧═══"*2}{"╩═══"}{"╧═══"*2}╝\n'
+        board_string = upper_lines
+        for index, line in enumerate(self.board):
+            row_list = []
+            for square_no, part in enumerate([line[:3], line[3:6], line[6:]], start=1):
+                row_square = '|'.join(str(item) for item in part)
+                row_list.extend(row_square)
+                if square_no != 3:
+                    row_list.append('║')
+
+            row = f'║ {" ".join(row_list)} ║\n'
+            row_empty = row.replace('0', ' ')
+            board_string += row_empty
+
+            if index < 8:
+                if index % 3 == 2:
+                    board_string += f'╠═══{"╪═══"*2}{"╬═══"}{"╪═══"*2}{"╬═══"}{"╪═══"*2}╣\n'
+                else:
+                    board_string += middle_lines
+            else:
+                board_string += lower_lines
+
+        return board_string
+
+    def find_empty_cell(self):
+        for row, contents in enumerate(self.board):
+            try:
+                col = contents.index(0)
+                return row, col
+            except ValueError:
+                pass
+        return None
+
+    def valid_in_row(self, row, num):
+        return num not in self.board[row]
+
+    def valid_in_col(self, col, num):
+        return all(
+            self.board[row][col] != num
+            for row in range(9)
+        )
+
+    def valid_in_square(self, row, col, num):
+        row_start = (row // 3) * 3
+        col_start=(col // 3) * 3
+        for row_no in range(row_start, row_start + 3):
+            for col_no in range(col_start, col_start + 3):
+                if self.board[row_no][col_no] == num:
+                    return False
+        return True
+
+    def is_valid(self, empty, num):
+        row, col = empty
+        valid_in_row = self.valid_in_row(row, num)
+        valid_in_col = self.valid_in_col(col, num)
+        valid_in_square = self.valid_in_square(row, col, num)
+        return all([valid_in_row, valid_in_col, valid_in_square])
+
+    def solver(self):
+        if (next_empty := self.find_empty_cell()) is None:
+            return True
+        else:
+            for guess in range(1, 10):
+                if self.is_valid(next_empty, guess):
+                    row, col = next_empty
+                    self.board[row][col] = guess
+                    if self.solver():
+                        return True
+                    self.board[row][col] = 0
+
+        return False
+
+def solve_sudoku(board):
+    gameboard = Board(board)
+--fcc-editable-region--
+    print('\nPuzzle to solve:')
```

Comment:
`print('\nPuzzle to solve:')` requires an additional step.

Reply:
I have left that intentionally in the seed code.